worker/worker-news-processor/index.js

```diff
@@ -558,116 +558,269 @@ async function processArticle(env, article, config) {
+  }
+  
+  // Step 6: Write article to KV and update ID index
+  const articleKey = `article:${nextArticleId}`;
+  await kv.put(articleKey, JSON.stringify(updatedArticle), {
+    expirationTtl: config.ID_INDEX_TTL
+  });
+  
+  // Update ID index
+  let idIndex = [];
+  try {
+    const idIndexData = await kv.get(config.KV_KEY_IDS, { type: 'json' });
+    if (idIndexData && Array.isArray(idIndexData)) {
+      idIndex = idIndexData;
+    }
+  } catch (error) {
+    // No index yet
+  }
+  
+  if (!idIndex.includes(nextArticleId)) {
+    idIndex.unshift(nextArticleId);
     
-    // Step 3: Process articles
-    const articlesToProcess = pendingArticles.slice(0, config.MAX_ARTICLES_PER_RUN);
-    console.log(`Processing ${articlesToProcess.length} articles this run...`);
+    // Limit index size
+    const maxArticles = config.MAX_STORED_ARTICLES || 500;
+    if (idIndex.length > maxArticles) {
+      idIndex = idIndex.slice(0, maxArticles);
+    }
     
-    let processedCount = 0;
+    await kv.put(
+      config.KV_KEY_IDS,
+      JSON.stringify(idIndex),
+      {
+        expirationTtl: config.ID_INDEX_TTL
+      }
+    );
+  }
+  
+  // Step 7: Update checkpoint based on success/failure
+  if (processingFailed) {
+    // Move article to try-later list
+    checkpoint.tryLater = (checkpoint.tryLater || []).filter(item => item.id !== nextArticleId);
+    checkpoint.tryLater.push({
+      id: nextArticleId,
+      article: updatedArticle,
+      failedAt: Date.now(),
+      reason: failureReason
+    });
     
-    // Process articles and collect updates
-    const updatePromises = [];
+    if (!checkpoint.processedIds.includes(nextArticleId)) {
+      checkpoint.processedIds.push(nextArticleId);
+    }
     
-    for (const item of articlesToProcess) {
-      const { id, article } = item;
-      console.log(`\nProcessing article ${processedCount + 1}/${articlesToProcess.length}: "${article.title?.substring(0, 50)}..."`);
-      
-      try {
-        // Process the article (one phase per run)
-        const updatedArticle = await processArticle(env, article, config);
-        
-        // Queue the write operation
-        const articleKey = `article:${id}`;
-        updatePromises.push(
-          env.CRYPTO_NEWS_CACHE.put(articleKey, JSON.stringify(updatedArticle), {
-            expirationTtl: config.ID_INDEX_TTL || 60 * 60 * 24 * 30
-          })
-        );
-        
-        processedCount++;
-      } catch (error) {
-        console.error(`  ✗ Error processing article:`, error);
-        // Continue with next article
+    checkpoint.currentArticleId = null;
+    checkpoint.currentArticle = null;
+  } else {
+    // Check if article is fully complete
+    const isComplete = 
+      !(updatedArticle.needsSentiment ?? true) && 
+      !(updatedArticle.needsSummary ?? true) && 
+      !(updatedArticle.contentTimeout && updatedArticle.contentTimeout < config.MAX_CONTENT_FETCH_ATTEMPTS);
```

Comment:
This code is repeated in multiple places.Just like `getArticleId` we should move this to a function.